sys/mips/gt64120.c

```diff
@@ -313,65 +305,74 @@ static int gt_pci_attach(device_t *pcib) {
   return bus_generic_probe(pcib);
 }
 
-static resource_t *gt_pci_alloc_resource(device_t *pcib, device_t *dev,
-                                         res_type_t type, int rid,
-                                         rman_addr_t start, rman_addr_t end,
-                                         size_t size, res_flags_t flags) {
+static resource_t *gt_pci_alloc_resource(device_t *dev, res_type_t type,
+                                         int rid, rman_addr_t start,
+                                         rman_addr_t end, size_t size,
+                                         res_flags_t flags) {
+  /* Currently all devices are logicaly attached to PCI bus,
+   * because we don't have PCI-ISA bridge implemented. */
+  assert(dev->bus == DEV_BUS_PCI && dev->parent->bus == DEV_BUS_PCI);
 
+  device_t *pcib = dev->parent;
   gt_pci_state_t *gtpci = pcib->state;
   bus_space_handle_t bh;
   rman_t *from = NULL;
 
-  /* Hack for ISA devices attached to PIIX4. TODO implement PCI-ISA bridge. */
-  if (type == RT_ISA) {
-    from = &gtpci->isa_io_rman;
-    bh = gtpci->isa_io->r_bus_handle;
+  if (type == RT_IOPORTS && end < 1024) {
+    /* Handle ISA device resources only. */
+    from = &gtpci->pci_io_rman;
+    bh = gtpci->pci_io->r_bus_handle;
   } else {
     /* Now handle only PCI devices. */
-
-    /* Currently all devices are logicaly attached to PCI bus, because we don't
-       have PCI-ISA bridge implemented. ISA devices are required to specify
-       RT_ISA_F flag, and have their dev->bus set to DEV_BUS_NONE. */
-    assert(dev->bus == DEV_BUS_PCI && dev->parent->bus == DEV_BUS_PCI);
-
-    /* Find identified bar by rid. */
     pci_device_t *pcid = pci_device_of(dev);
+    /* Find identified bar by rid. */
     pci_bar_t *bar = &pcid->bar[rid];
 
     if (bar->size == 0)
       return NULL;
 
+    /* The size of allocated resource must not be larger than what was reported
+     * by corresponding BAR. */
+    size = min(size, bar->size);
+
     if (type == RT_MEMORY) {
       from = &gtpci->pci_mem_rman;
-      size = bar->size;
       bh = gtpci->pci_mem->r_start;
     } else if (type == RT_IOPORTS) {
       from = &gtpci->pci_io_rman;
-      size = bar->size;
       bh = gtpci->pci_io->r_bus_handle;
     } else {
```

Comment:
Shouldn't GT support `RT_IRQ`?

Reply:
Let's postpone irq management to another PR. This PR is already bloated.